scripts/lib/releaseNotes.ts

```diff
@@ -150,3 +151,140 @@ function addNewReleaseNoteToc(releaseNotesNode: any, newVersion: string) {
     });
   }
 }
+
+/**
+ * Sorts the release notes dictionary by the patch version
+ */
+export function sortReleaseNotesVersions(markdownByPatchVersion: {
+  [id: string]: string;
+}) {
+  const markdownByPathEntries = Object.entries(markdownByPatchVersion);
+
+  const markdownByPatchSortedArray = markdownByPathEntries.sort(
+    ([version1], [version2]) => {
+      const versionPatch1 = version1.split("rc").slice(0, 1)[0];
+      const versionPatch2 = version2.split("rc").slice(0, 1)[0];
+      const comparison = version1.localeCompare(version2);
+
+      if (versionPatch1 == versionPatch2) {
+        // The release candidates should appear first
+        if (version1.length < version2.length) {
+          return 1;
+        } else if (version1.length > version2.length) {
+          return -1;
+        }
+      }
+      return comparison;
+    },
+  );
+
+  return Object.fromEntries(markdownByPatchSortedArray);
+}
+
+/**
+ * Process the markdown dividing it into small sections for each patch version
+ *
+ * Returns a tuple:
+ *   1. A Set of strings representing all the minor versions we found in the
+ *      markdown.
+ *   2. A dictionary with the corresponding markdown for each patch version
+ *      found.
+ */
+export function extractMarkdownReleaseNotesPatches(
+  markdown: string,
+): [Set<string>, { [id: string]: string }] {
+  const sectionsSplit = markdown.split(/\n## (?=[0-9])/);
+  const sections: string[] = sectionsSplit.slice(1, sectionsSplit.length);
+
+  const minorVersionsFound = new Set<string>();
+  const markdownByPatchVersion: { [id: string]: string } = {};
+
+  sections.forEach((section) => {
+    const versionPatch = section.split("\n").slice(0, 1)[0];
+    const versionMinor = versionPatch.split(".").slice(0, 2).join(".");
+    minorVersionsFound.add(versionMinor);
+
+    const content = section.split("\n");
+    content.shift();
+    markdownByPatchVersion[versionPatch] = `## ${versionPatch}\n${content.join(
+      "\n",
+    )}`;
+  });
+
+  return [minorVersionsFound, markdownByPatchVersion];
+}
+
+/**
+ * Updates the release notes files by adding the notes to their corresponding version
+ * file.
+ */
+export async function writeReleaseNotes(pkg: Pkg, releaseNoteMarkdown: string) {
+  // Dictionary to store the file header in case we have to reconstruct a historical file
+  const FilesHeaders: { [id: string]: string } = {};
+  const basePath = `${getRoot()}/docs/api/${pkg.name}/release-notes/`;
+
+  const [minorVersionsFound, markdownByPatchVersion] =
+    extractMarkdownReleaseNotesPatches(releaseNoteMarkdown);
+
+  // Read the current release notes for each version found
+  for (let version of minorVersionsFound) {
+    const versionPath = `${basePath}${version}.md`;
+
+    if (!(await pathExists(versionPath))) {
+      // We don't have any release note file for that version
+      continue;
+    }
```

Comment:
This is an edge case possible if, for some reason, we generate two versions at the same time. For example, if we add a version using the historical argument, the release notes will not be created, and when we add the current version, the file for the former won't exist.

This shouldn't happen often, but if someday we have to regenerate several versions, for some reason, the script will handle the case correctly.

Reply:
Oh - I think this probably speaks to my other comment about realizing I think we have a bug how `--historical` stops us from generating release notes?

Reply:
I continue to be confused about this point. 

> we generate two versions at the same time.

That's not technically possible. The script only can generate one release at a time. 

> For example, if we add a version using the historical argument, the release notes will not be created, and when we add the current version, the file for the former won't exist.

I don't think I'm understanding. If the historical arg has 0.44 and 0.45, then we should expect both those release note files to exist. Now we generate current, which has 0.44, 0.45, and 0.46: it should generate all three of those files.

Reply:
> That's not technically possible.

Yes, sorry, I meant in the same PR. In a situation where you skipped one version, and you want now to add the current and the previous you skipped.

> I don't think I'm understanding.

The problem that I wanted to describe can't happen anymore because now, the script generates release notes even when we use the historical argument. However, the problem is still there.

When you generate a version for the first time, the release notes for that new version are not present in the release notes folder. The script will read the release notes in order to find old versions we modified, and therefore we need to update from the files in the folder. To update the old versions we need to read the current release notes for those versions. The catch is that the script will try to read the current version as well, but that file doesn't exist yet, so we need to prevent it from breaking.

We need to read the current release notes file because we don't want to overwrite the whole file we had. There are some parts like the front matter or text added manually (like the table in qiskit v0.44) that we want to preserve. In addition, if we don't modify a version, we want to have the notes and not lose them (not sure if this can happen because of the way Qiskit generates the release notes now though).

Reply:
> That's not technically possible.

Yes, sorry, I meant in the same PR. In a situation where you skipped one version, and you want now to add the current and the previous you skipped as historical without having it generated as current before.

> I don't think I'm understanding.

The problem that I wanted to describe can't happen anymore because now, the script generates release notes even when we use the historical argument. However, the problem is still there.

When you generate a version for the first time, the release notes for that new version are not present in the release notes folder. The script will read the release notes (html converted to md) in order to find old versions we modified, and we need to update from the files in the folder. To update the old versions, we need to read the current release notes for those versions. The catch is that the script will try to read the current version as well, but the file doesn't exist yet, so we need to prevent it from breaking.

We need to read the current release notes files because we don't want to overwrite everything we had. There are some parts like the front matter or text added manually (like the table in qiskit v0.44 or a custom description) that we want to preserve. Your comment made me realize that we don't need to go through each section because the release notes should contain all the versions by default, so we can just overwrite that part with the new content. In the last commit, I removed some unnecessary code. Thank you Eric.